builtin/logical/aws/path_roles.go

```diff
@@ -58,94 +85,342 @@ func pathRoles() *framework.Path {
 }
 
 func (b *backend) pathRoleList(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
-	entries, err := req.Storage.List(ctx, "policy/")
+	b.roleMutex.RLock()
+	defer b.roleMutex.RUnlock()
+	entries, err := req.Storage.List(ctx, "role/")
 	if err != nil {
 		return nil, err
 	}
-	return logical.ListResponse(entries), nil
-}
-
-func pathRolesDelete(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
-	err := req.Storage.Delete(ctx, "policy/"+d.Get("name").(string))
+	legacyEntries, err := req.Storage.List(ctx, "policy/")
 	if err != nil {
 		return nil, err
 	}
 
+	return logical.ListResponse(append(entries, legacyEntries...)), nil
+}
+
+func (b *backend) pathRolesDelete(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	for _, prefix := range []string{"policy/", "role/"} {
+		err := req.Storage.Delete(ctx, prefix+d.Get("name").(string))
+		if err != nil {
+			return nil, err
+		}
+	}
+
 	return nil, nil
 }
 
-func pathRolesRead(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
-	entry, err := req.Storage.Get(ctx, "policy/"+d.Get("name").(string))
+func (b *backend) pathRolesRead(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	entry, err := b.lockedRoleRead(ctx, req.Storage, d.Get("name").(string))
 	if err != nil {
 		return nil, err
 	}
 	if entry == nil {
 		return nil, nil
 	}
 
-	val := string(entry.Value)
-	if strings.HasPrefix(val, "arn:") {
-		return &logical.Response{
-			Data: map[string]interface{}{
-				"arn": val,
-			},
-		}, nil
-	}
 	return &logical.Response{
-		Data: map[string]interface{}{
-			"policy": val,
-		},
+		Data: entry.toResponseData(),
 	}, nil
 }
 
-func useInlinePolicy(d *framework.FieldData) (bool, error) {
-	bp := d.Get("policy").(string) != ""
-	ba := d.Get("arn").(string) != ""
+func legacyRoleData(d *framework.FieldData) (string, error) {
+	policy := d.Get("policy").(string)
+	arn := d.Get("arn").(string)
+
+	switch {
+	case policy == "" && arn == "":
+		return "", nil
+	case policy != "" && arn != "":
+		return "", errors.New("only one of policy or arn should be provided")
+	case policy != "":
+		return policy, nil
+	default:
+		return arn, nil
+	}
+}
+
+func (b *backend) pathRolesWrite(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	var resp logical.Response
+
+	roleName := d.Get("name").(string)
+	if roleName == "" {
+		return logical.ErrorResponse("missing role name"), nil
+	}
+
+	b.roleMutex.Lock()
```

Comment:
The reason for the mutex is to prevent race conditions when multiple write entries are made at the same time on the HTTP endpoint. Or is this something I don't need to worry about?

Reply:
We don't guarantee order of requests and the upgrade itself should lock, so it is safe to not worry about locks for the http requests.

Reply:
Do you mean the upgrade of the role entry from legacy to current should take the lock? I still don't think that's safe and we still need the lock. Imagine the following scenario:

1. Request 1 comes in to change the `policy_document` paramemter of an existing, already-upgraded role
1. Request 2 comes in at the same time to change the role_arns of the same role.
1. Request 1 and Request 2 both read the same, old role data
1. Request 1 and Request 2 both modify the same role in different ways
1. Request 1 and Request 2 will race to write different role data to the storage backend, and the loser of the race will have the request lost

If the semantics of the write involved writing the entirety of the role at once, it would be fine. But since the semantics also allow modifying only some parameters of the role in place, we need a write lock here.

Or am I missing something else?

Reply:
I think it's fine...I don't anticipate that this is going to be called a ton so contention shouldn't be bad.